internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,318 @@
+type standardDynamicScaleDownProcessor struct {
+	clusterID                    string
+	regionsSupportedInstanceType config.InstanceTypeMap
+	// kafkaStreamingUnitCountPerClusterList must not contain any element
+	// with a Status attribute with value 'failed'
+	kafkaStreamingUnitCountPerClusterList services.KafkaStreamingUnitCountPerClusterList
+	// indexesOfStreamingUnitForSameClusterID is the index of the straming unit in the "kafkaStreamingUnitCountPerClusterList" that has the
+	// same "clusterID" as the one that's current being processed
+	indexesOfStreamingUnitForSameClusterID []int
+	supportedKafkaInstanceTypesConfig      *config.SupportedKafkaInstanceTypesConfig
+	clusterService                         services.ClusterService
+
+	// dryRun controls whether the ScaleDown method performs real actions.
+	// Useful when you don't want to trigger a real scale down.
+	dryRun bool
+}
+
+var _ dynamicScaleDownEvaluator = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownExecutor = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{}
+
+// ShouldScaleDown indicates whether a new data plane
+// cluster should be created for a given instance type in the given provider
+// and region.
+// It returns true if all the following conditions happen:
+// 1. If specified the cluster is empty i.e it does not contain any streaming unit
+// 2. If the cluster can be removed without triggering a scale up action
+// Otherwise false is returned.
+// Note: This method assumes kafkaStreamingUnitCountPerClusterList does not
+// contain elements with the Status attribute with the 'failed' value.
+// Thus, if the type is constructed with the assumptions being true, it
+// can be considered as the 'failed' state Clusters are not included in
+// the calculations.
+func (p *standardDynamicScaleDownProcessor) ShouldScaleDown() (bool, error) {
+	// First let's check if the cluster is empty
+	for _, i := range p.indexesOfStreamingUnitForSameClusterID {
+		clusterIsNotEmpty := p.kafkaStreamingUnitCountPerClusterList[i].Count > 0
+		if clusterIsNotEmpty {
+			return false, nil
+		}
+	}
+
+	// let's check if the cluster can be safely removed without causing a scale up event
+	if len(p.regionsSupportedInstanceType) == 0 { // if no region limits are available it means that this cluster is in a region that's not supported anymore, we can safely delete it if it is empty
+		return true, nil
+	}
+
+	newkafkaStreamingUnitCountPerClusterList := services.KafkaStreamingUnitCountPerClusterList{}
+
+	clusterStatesTowardReadyState := []string{
+		api.ClusterProvisioning.String(), api.ClusterProvisioned.String(),
+		api.ClusterAccepted.String(), api.ClusterWaitingForKasFleetShardOperator.String(),
+	}
+
+	for _, suCount := range p.kafkaStreamingUnitCountPerClusterList {
+		// Ignore clusters in terraforming states for scale up evaluation as they are not ready yet.
```

Comment:
Is ignoring this correct?

The ScaleUp processor uses that information to check if there is an ongoing scale up action in progress so we should keep them.

Reply:
They are also used to perform capacity calculations

Reply:
Before deleting a cluster, we want to be sure that there is a sibling cluster in a ready state i.e capable of accepting a kafka. If we end up considering the clusters that are not ready yet, we might delete a cluster and find ourselves having to wait until the other clusters that are not ready yet to be ready to start provisioning Kafkas on them.

The approach taken here is to be conservative and only evaluate if scale up needs to be performed against sibling cluster that are "ready". 

To summarise:

If we've cluster `A` that's ready and empty and cluster `B` that's in `accepted`.
We do not want to delete cluster `A` up until we are sure that we've one sibling cluster that's ready i.e until cluster `B` becomes ready.

Reply:
Can't we just check that before?
I am finding this quite confusing/fragile as it has lots of implicit behavior around scale up manager and it assumes clusters toward ready state are there to evaluate some of the conditions.

Reply:
> Can't we just check that before?

Can you expand more on this? You want these clusters filtered out of the list completely at the start of scale down evaluation?

> scale up manager assumes clusters toward ready state are there to evaluate some of the conditions

the scale up evaluator is doing it is job correctly for scale. 

But for dynamic scaling down, we do not want to rely only on the "mare" presence of a sibling cluster for us to delete a candidate cluster. The sibling cluster has to be in `ready` state hence this filter: this ensures that no any false signaling that "scale up won't be triggered" will be thrown by the scale up evaluation after the candidate cluster is deleted.

Reply:
> Can you expand more on this? You want these clusters filtered out of the list completely at the start of scale down evaluation?

Check if there is another cluster in Ready state in that region. Then we don't need to call the scale up mgr at all. It is clear we won't want to delete it

Reply:
My concern is that dynamic scale up mgr uses clusters toward ready state to calculate different conditions and here we are excluding them so those capacities will be ignored which they shouldn't

Reply:
> Check if there is another cluster in Ready state in that region. That way we know we won't want to delete it if it's the only cluster

Ah right. This is really something that already handled nicely in the `.ShouldScaleUp()`.  Hence the re-use of that behavior to ensure that when we consider that only ready clusters in the region, then no scale up will need to be performed in that region if a candidate cluster is removed.

Reply:
All the information are provided, except for the clusters in terraforming phase which are filtered out intentionally.
That's because, we don't want https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/8aa0a7614ce607c101f0ef85533fb8f9684c7b13/internal/kafka/internal/workers/cluster_mgrs/dynamic_scaleup_mgr.go#L254-L257 to report `no need to scale up after the candidate cluster is removed because we've an ongoing scale up action.` Because if that happens, it might create a situation where we do not have a ready cluster in a region because we have just wrongly deleted the only left

Reply:
The information of those clusters is not only taken into account for that condition but also for other ones

Reply:
The other part where they are used is in capacity calculations? if so, I think `.Count` and `.MaxUnits` will be `0` for these clusters until they are `ready: https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/8aa0a7614ce607c101f0ef85533fb8f9684c7b13/internal/kafka/internal/services/data_plane_cluster.go#L89-L101. 
So even though the information are considered, they do not add/remove much.

Reply:
:thinking: that's correct I think.
Can we expand the documentation around all of this? I think it is not trivial to understand what it is being done and why.
For example the current doc says that it would scale up if:
`// 2. If the cluster can be removed without triggering a scale up action`

But the triggering of a scale up action is really modified as the provided list of clusters is altered to kind of alter what is being returned by shouldScaleUp.

Reply:
That's true. 

This doc https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/docs/architecture/data-plane-osd-cluster-dynamic-scaling.md#osd-cluster-deletion-evaluation is up to date. I think only the method `ShouldScaleDown()` description  is to be updated.